subsys/bluetooth/audio/ascs.c

```diff
@@ -285,198 +398,143 @@ void ascs_ep_set_state(struct bt_bap_ep *ep, uint8_t state)
+		switch (old_state) {
+		case BT_BAP_EP_STATE_QOS_CONFIGURED:
+		case BT_BAP_EP_STATE_ENABLING:
 			break;
+		default:
+			goto invalid_transition;
+		}
+
+		break;
+	case BT_BAP_EP_STATE_STREAMING:
+		switch (old_state) {
+		case BT_BAP_EP_STATE_ENABLING:
 		case BT_BAP_EP_STATE_STREAMING:
-			switch (old_state) {
-			case BT_BAP_EP_STATE_ENABLING:
-			case BT_BAP_EP_STATE_STREAMING:
-				break;
-			default:
-				BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-					      bt_bap_ep_state_str(old_state),
-					      bt_bap_ep_state_str(ep->status.state));
-				return;
-			}
+			break;
+		default:
+			goto invalid_transition;
+		}
 
-			if (state_changed && ops->started != NULL) {
-				ops->started(stream);
-			} else if (!state_changed && ops->metadata_updated) {
-				ops->metadata_updated(stream);
+		break;
+	case BT_BAP_EP_STATE_DISABLING:
+		switch (old_state) {
+		case BT_BAP_EP_STATE_ENABLING:
+		case BT_BAP_EP_STATE_STREAMING:
+			if (ase->ep.dir != BT_AUDIO_DIR_SOURCE) {
+				goto invalid_transition;
 			}
+			break;
+		default:
+			goto invalid_transition;
+		}
 
+		break;
+	case BT_BAP_EP_STATE_RELEASING:
+		switch (old_state) {
+		case BT_BAP_EP_STATE_CODEC_CONFIGURED:
+		case BT_BAP_EP_STATE_QOS_CONFIGURED:
+		case BT_BAP_EP_STATE_ENABLING:
+		case BT_BAP_EP_STATE_STREAMING:
 			break;
 		case BT_BAP_EP_STATE_DISABLING:
-			if (ep->dir == BT_AUDIO_DIR_SOURCE) {
-				switch (old_state) {
-				case BT_BAP_EP_STATE_ENABLING:
-				case BT_BAP_EP_STATE_STREAMING:
-					break;
-				default:
-					BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-						      bt_bap_ep_state_str(old_state),
-						      bt_bap_ep_state_str(ep->status.state));
-					return;
-				}
-			} else {
-				/* Sinks cannot go into the disabling state */
-				BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-					      bt_bap_ep_state_str(old_state),
-					      bt_bap_ep_state_str(ep->status.state));
-				return;
+			if (ase->ep.dir != BT_AUDIO_DIR_SOURCE) {
+				goto invalid_transition;
 			}
+			break;
+		default:
+			goto invalid_transition;
+		}
 
-			ep->receiver_ready = false;
-
-			if (ops->disabled != NULL) {
-				ops->disabled(stream);
-			}
+		break;
+	default:
+		goto invalid_transition;
+	}
 
-			break;
-		case BT_BAP_EP_STATE_RELEASING:
-			switch (old_state) {
-			case BT_BAP_EP_STATE_CODEC_CONFIGURED:
-			case BT_BAP_EP_STATE_QOS_CONFIGURED:
-			case BT_BAP_EP_STATE_ENABLING:
-			case BT_BAP_EP_STATE_STREAMING:
-				break;
-			case BT_BAP_EP_STATE_DISABLING:
-				if (ep->dir == BT_AUDIO_DIR_SOURCE) {
-					break;
-				} /* else fall through for sink */
-
-				/* fall through */
-			default:
-				BT_ASSERT_MSG(false, "Invalid state transition: %s -> %s",
-					      bt_bap_ep_state_str(old_state),
-					      bt_bap_ep_state_str(ep->status.state));
-				return;
-			}
+	ase->state_pending = state;
 
-			ep->receiver_ready = false;
+	err = k_work_submit(&ase->state_transition_work);
+	if (err < 0) {
+		return err;
```

Comment:
Maybe we should log this? 

If this indeed fails, then we can't really recover from it, can we?

Reply:
Indeed. We can't really do much besides logging this as an `LOG_ERR`; I don't think we can do anything to recover from it